### PR TITLE
Pretty printer fixes

### PR DESCRIPTION
**Pull request by OceanOak**
No changelog

This PR addresses a couple pretty printer issues

### Line breaks and spaces
Issue:
```
  let indent (s: String) : String =
        s
        |> String.split "
          
          "
       |> List.map ((fun line -> 
               match line with 
              |  ->
                ""
              | line ->
                ("  ") ++ (line))) 
            |> String.join ("
            
            ")
```

Fixed:
```
 let indent (s: String) : String =
      s 
      |> String.split ("\n")
      |> List.map ((fun line -> 
         match line with 
        | "" ->
          ""
        | line ->
          ("  ") ++ (line))) 
      |> String.join ("\n")
```

### Floats
Issue:
```
{ model = "gpt-3.5-turbo"; max_tokens = 700; temperature = 0.6999999999999999555910790149937383830547332763671875 
````

Fixed:
```
{ model = "gpt-3.5-turbo"; max_tokens = 700; temperature = 0.7}
```

I've also tested these, and they are being displayed correctly
let f = 1.0
let g = 0.799999
let s = 0.7
let d = 2.2
let y = 20.33

### CODE REVIEW

File: packages/darklang/prettyPrinter/programTypes.dark [hunk 1]
```diff
@@ -458,21 +455,31 @@ module Darklang =
...
-          $"{signPart}{wholePart}{remainderPart}"
 
         | EChar(_id, c) -> "\"" ++ c ++ "\""
 
         | EString(_id, segments) ->
           match segments with
           | [] -> "\"\""
-          | [ StringText text ] -> $"\"{text}\""
+          | [ StringText text ] ->
+            let escapedText = String.replaceAll text "\n" "\\n"
```
Comment (OceanOak) [inline packages/darklang/prettyPrinter/programTypes.dark:481]: I used replaceAll in this case since I couldn't use String.map  (wasn't allowed to replace a char with a string). 
Do you have any suggestions for a more efficient approach that would allow checking for things other than line breaks, such as spaces?

Comment (pbiggar) [reply packages/darklang/prettyPrinter/programTypes.dark:481]: What about something like:

`String.split ... |> List.map (replace ...) |> String.concat`

File: packages/darklang/prettyPrinter/programTypes.dark [hunk 2]
```diff
@@ -458,21 +455,31 @@ module Darklang =
...
+              match rem with
+              | Ok r ->
+                  if r == 0.0 then ""
+                  else (r  |> Float.toString) |> String.dropFirst 2
+              | Error e -> e
+          let wholePart =
+              match whole with
+              | "0" -> if remainderPart != "" then "0" else ""
+              | _ -> whole
+          let formattedFloat =
```
Comment (pbiggar) [inline packages/darklang/prettyPrinter/programTypes.dark:469]: I'm finding it hard to understand what's happening here. Could you add comments (with examples maybe) to discuss the edge cases you're handling?

Also can you add tests? I think you should be able to add tests within backend/testfiles/execution.

File: packages/darklang/prettyPrinter/programTypes.dark [hunk 3]
```diff
@@ -300,12 +300,9 @@ module Darklang =
         | MPBool(_id, b) -> Bool.toString b
         | MPInt(_id, i) -> Int.toString i
         | MPFloat(_id, sign, whole, remainder) ->
```
[the remark this answers is not in the document]
Comment (pbiggar) [reply packages/darklang/prettyPrinter/programTypes.dark:302]: Yes I think those would be the same. Maybe extract the function to do both.

File: packages/darklang/prettyPrinter/programTypes.dark (hunk 1 above)
Comment (pbiggar) [reply packages/darklang/prettyPrinter/programTypes.dark:481]: That said, `String.replaceAll` might be fine. You could do something like:

```
text
|> String.replaceAll "\n" "\\n"
|> String.replaceAll "\t" "\\t"
|> String.replaceAll "\r" "\\r"
```

**Review by pbiggar — CHANGES_REQUESTED** [review]

File: packages/darklang/prettyPrinter/programTypes.dark (hunk 2 above)
Comment (OceanOak) [reply packages/darklang/prettyPrinter/programTypes.dark]: We don't need this anymore. it was used when the number lies between 0 and 1, to keep the "0" in whole part.

File: packages/darklang/prettyPrinter/common.dark [hunk 4]
```diff
@@ -16,4 +16,21 @@ module Darklang =
     let sign (s: PACKAGE.Darklang.Stdlib.Sign) : String =
       match s with
       | Positive -> ""
-      | Negative -> "-"
\ No newline at end of file
+      | Negative -> "-"
+
+    let escapeSpecialCharacters (s: String) : String =
```
Comment (OceanOak) [inline packages/darklang/prettyPrinter/common.dark:21]: Is this file the right place for these functions?

File: packages/darklang/prettyPrinter/common.dark [hunk 5]
```diff
@@ -16,4 +16,21 @@ module Darklang =
...
+    let processRemainder (remainder: String) : String =
+      let rem = $".{remainder}" |> Float.parse
+
+      match rem with
+      | Ok r ->
+        if r == 0.0 then
+          "0"
+        else
+          (r |> Float.toString) |> String.dropFirst 2
+      | Error e -> e
```
Comment (OceanOak) [inline packages/darklang/prettyPrinter/common.dark:36]: Is it bad to return the error here?

File: packages/darklang/prettyPrinter/common.dark (hunk 4 above)
Comment (pbiggar) [reply packages/darklang/prettyPrinter/common.dark:21]: Sure.

File: packages/darklang/prettyPrinter/common.dark (hunk 5 above)
Comment (pbiggar) [reply packages/darklang/prettyPrinter/common.dark:32]: what would you expect to happen if you parse an invalid float?

Comment (OceanOak) [reply packages/darklang/prettyPrinter/common.dark:32]: I expect an error or an exception to be thrown indicating that parsing an invalid float isn't possible
The issue is if I have something like 
let z = 0.abc
let a = 3.14abc
in the pretty printer is is displayed as 
let z = 0
let a = 0

Comment (pbiggar) [reply packages/darklang/prettyPrinter/common.dark:32]: The difference here, if I understand it correctly, is that this is pretty printing EFloats, where it should not be possible to have invalid floats present.

Comment (OceanOak) [reply packages/darklang/prettyPrinter/common.dark:32]: Currently, when we send an invalid float like `1.b`, it doesn't reach the `EFloat` branch; instead, it is considered as an `EInt`, and as a result, returns 0. I don't believe this issue is related to the pretty printer. I've added some tests in 'Int.test' for this scenario (i.e., 'Int.toString 1.ab = "0"' passes)
